jormungandr/src/blockchain/checkpoints.rs

```diff
@@ -0,0 +1,70 @@
+use crate::{blockcfg::HeaderHash, blockchain::Ref};
+use std::sync::Arc;
+
+/// list of pre-computed checkpoints from a given [`Ref`].
+///
+/// [`Ref`]: ./struct.Ref.html
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Checkpoints(Vec<HeaderHash>);
+
+impl Checkpoints {
+    /// create a new list of checkpoints from the given starting point (tip).
+    ///
+    /// The function make use of the `Ref` object pointing to previous epoch in order
+    /// to populate _interestingly_ spaced checkpoints.
+    ///
+    /// For now the algorithm provide the current block, the parent, the last block of the
+    /// previous epoch, the last block of the epoch before that... until the block0.
+    pub fn new_from(from: Arc<Ref>) -> Self {
+        let mut checkpoints = vec![from.hash(), from.block_parent_hash().clone()];
+
+        let mut ignore_prev = 0;
+        let mut current_ref = from;
+        while let Some(prev_epoch) = current_ref.last_ref_previous_epoch() {
```

Comment:
If the starting block's parent is the last in the previous epoch, will it be added twice?

Reply:
good catch!